packages/firestore/src/util/async_queue.ts

```diff
@@ -190,11 +190,20 @@ export class AsyncQueue {
   // assertion sanity-checks.
   private operationInProgress = false;
 
+  /**
+   * Adds a new operation to the queue without waiting for it to complete (i.e.
+   * we ignore the Promise result).
+   */
+  enqueue<T extends Unknown>(op: () => Promise<T>): void {
+    // tslint:disable-next-line:no-floating-promises
+    this.enqueueAndWait(op);
+  }
+
   /**
    * Adds a new operation to the queue. Returns a promise that will be resolved
    * when the promise returned by the new operation is (with its value).
    */
-  enqueue<T extends Unknown>(op: () => Promise<T>): Promise<T> {
+  enqueueAndWait<T extends Unknown>(op: () => Promise<T>): Promise<T> {
```

Comment:
The common case (and thus most likely to have a parallel on the other platforms) would be to enqueue without dropping the promise, no?

If so I'd prefer we continue to name this enqueue and call the somewhat danergous enqueue and drop promise something else.

Reply:
Actually there is no enqueueAndWait() on the other platforms (dispatchAsync on iOS and enqueue on Android both return void and don't block).  Instead you have to monkey with completions / TaskCompletionSource to plumb the async operation result back to the caller.

enableNetwork on web:
```
  enableNetwork(): Promise<void> {
    return this.asyncQueue.enqueueAndWait(() => {
      return this.remoteStore.enableNetwork();
    });
  }
```

enableNetwork on iOS:
```
- (void)enableNetworkWithCompletion:(nullable FSTVoidErrorBlock)completion {
  [self.workerDispatchQueue dispatchAsync:^{
    [self.remoteStore enableNetwork];
    if (completion) {
      self->_userExecutor->Execute([=] { completion(nil); });
    }
  }];
}
```

enableNetwork on Android:
```
  public Task<Void> enableNetwork() {
    final TaskCompletionSource<Void> source = new TaskCompletionSource<>();
    asyncQueue.enqueue(
        () -> {
          remoteStore.enableNetwork();
          source.setResult(null);
        });
    return source.getTask();
  }
```

So I think enqueue() is arguably the common case and is named consistently with the other platforms.

Also, I can't think of a good name if I make enqueue() the "blocking" one.  enqueueAndForget()?

Reply:
I guess I consider enqueue-that-returns-a-promise on web to be equivalent to enqueue-that-plumbs-through-callbacks on iOS and equivalent to enqueue-with-a-manually-wrapping TaskCompletionSource on Android.

We could have Android's enqueue do the TaskCompletionSource monkeying as well (TaskCompletionSource being the equivalent of a Promise), which would bring it further into line with web.

In any case, I think we generally want to tie actions enqueued with a user-visible promise so this should be be the shorter, more obvious name.

Regarding naming: enqueueAndWait suggests actually waiting, which isn't what's happening. Meanwhile enqueueAndForget seems to reasonably suggest what it's doing to me.

Reply:
Fair enough.  Renamed!

I'll take a look at the Android change.